src/zenml/steps/utils.py

```diff
@@ -38,57 +131,183 @@ def resolve_type_annotation(obj: Any) -> Any:
+            return
+
+        self._inside_function = True
+        self.generic_visit(node)
+
+    visit_FunctionDef = _visit_function
+    visit_AsyncFunctionDef = _visit_function
+
+
+class OnlyNoneReturnsVisitor(ReturnVisitor):
+    """Checks whether a function AST contains only `None` returns."""
+
+    def __init__(self) -> None:
+        """Initializes a visitor instance."""
+        super().__init__()
+        self.has_only_none_returns = True
+
+    def visit_Return(self, node: ast.Return) -> None:
+        """Visit a return statement.
+
+        Args:
+            node: The return statement to visit.
+        """
+        if node.value is not None:
+            if isinstance(node.value, (ast.Constant, ast.NameConstant)):
+                if node.value.value is None:
+                    return
+
+            self.has_only_none_returns = False
+
+
+class TupleReturnVisitor(ReturnVisitor):
+    """Checks whether a function AST contains tuple returns."""
+
+    def __init__(self) -> None:
+        """Initializes a visitor instance."""
+        super().__init__()
+        self.has_tuple_return = False
+
+    def visit_Return(self, node: ast.Return) -> None:
+        """Visit a return statement.
+
+        Args:
+            node: The return statement to visit.
+        """
+        if isinstance(node.value, ast.Tuple) and len(node.value.elts) > 1:
+            self.has_tuple_return = True
+
+
+def has_tuple_return(func: Callable[..., Any]) -> bool:
+    """Checks whether a function returns multiple values.
+
+    Multiple values means that the `return` statement is followed by a tuple
+    (with or without brackets).
 
     Example:
-        `get_args(Union[int, str]) == (int, str)`
+    ```
+    def f1():
+      return 1, 2
+
+    def f2():
+      return (1, 2)
+
+    def f3():
+      var = (1, 2)
+      return var
+
+    has_tuple_return(f1)  # True
+    has_tuple_return(f2)  # True
+    has_tuple_return(f3)  # False
+    ```
 
     Args:
-        obj: The annotation.
+        func: The function to check.
 
     Returns:
-        The args of the Union annotation.
+        Whether the function returns multiple values.
     """
-    return tuple(
-        pydantic_typing.get_origin(v) or v
-        for v in pydantic_typing.get_args(obj)
-    )
+    source = textwrap.dedent(source_code_utils.get_source_code(func))
+    tree = ast.parse(source)
+
+    visitor = TupleReturnVisitor()
+    visitor.visit(tree)
+
+    return visitor.has_tuple_return
 
 
-def parse_return_type_annotations(return_annotation: Any) -> Dict[str, Any]:
-    """Parse the returns of a step function into a dict of resolved types.
+def has_only_none_returns(func: Callable[..., Any]) -> bool:
+    """Checks whether a function contains only `None` returns.
 
-    Called within `BaseStepMeta.__new__()` to define `cls.OUTPUT_SIGNATURE`.
+    A `None` return could be either an explicit `return None` or an empty
+    `return` statement.
```

Comment:
Maybe also here an example like above could make this clearer

Reply:
Added one